### DISCUSSION TITLE
Proper way to clear out an inline-keyboard?

### DISCUSSION THREAD

**Discussion by Bottleface** [windkh/node-red-contrib-telegrambot#354] [Q&A] [answered]
I've tried every way I can think to clear out an inline-keyboard without throwing an error.  I've tried many variations of sending a _blank_  `reply_markup`, the thing is that most ways do in fact remove the keyboard, they just also throw the following error:

```
Caught exception in sender node:
Error: ETELEGRAM: 400 Bad Request: object expected as reply markup
when processing message: ...
```

I've also tried both the `editReplyMarkup` and `editMessage` calls and both behave the same.  Here are the variations I've tried:

```
msg.payload = {
    "type":"editMessageReplyMarkup",
    "content": "-",
    "options": {
        "chat_id":  msg.payload.chatId,
        "message_id": msg.payload.messageId,
        "reply_markup": null
    }
};
```
***
```
msg.payload = {
    "type":"editMessageReplyMarkup",
    "content": "-",
    "options": {
        "chat_id":  msg.payload.chatId,
        "message_id": msg.payload.messageId,
        "reply_markup": ""
    }
};
```
***
```
msg.payload = {
    "type":"editMessageReplyMarkup",
    "content": "-",
    "options": {
        "chat_id":  msg.payload.chatId,
        "message_id": msg.payload.messageId,
        "reply_markup": '{"inline_keyboard":[]}'
    }
};
```
***
```
msg.payload = {
    "type":"editMessageReplyMarkup",
    "content": "-",
    "options": {
        "chat_id":  msg.payload.chatId,
        "message_id": msg.payload.messageId
    }
};
```
***

Does anyone know how to clear out an inline-keyboard properly without throwing errors?

**Comment by windkh** [comment, marked as answer]
_(Drafted by Claude, Anthropic's AI coding assistant, after walking through your variations against the Telegram Bot API spec and the node-telegram-bot-api signature. Posted by the maintainer.)_

@Bottleface — apologies for the very late reply. Short answer: the variations you tried put the markup in the wrong field of the payload. The sender node maps `msg.payload.content` to the **first positional argument** of `editMessageReplyMarkup`, which is the markup object itself — not a placeholder string. Putting a `"-"` (or any string) there is what triggers `400 Bad Request: object expected as reply markup`.

### Why it failed

Looking at [out-node.js](https://github.com/windkh/node-red-contrib-telegrambot/blob/master/telegrambot/nodes/out-node.js) for the `editMessageReplyMarkup` branch:

```js
telegramBot[type](msg.payload.content, msg.payload.options || {})
```

It calls `telegramBot.editMessageReplyMarkup(content, options)`, and [node-telegram-bot-api's signature](https://github.com/yagop/node-telegram-bot-api/blob/master/doc/api.md#editMessageReplyMarkup) for that method is:

```js
editMessageReplyMarkup(replyMarkup, options)
```

So **the first argument *is* the reply markup**, not the message text. Your `content: "-"` was being passed to Telegram as the literal reply markup, and Telegram rejected the string with "object expected as reply markup". The `reply_markup` field you set inside `options` was extra noise on top.

### The working shape

```js
msg.payload = {
    type: "editMessageReplyMarkup",
    content: { inline_keyboard: [] },    // ← the markup object goes HERE, not in options
    options: {
        chat_id: msg.payload.chatId,
        message_id: msg.payload.messageId
    }
};
return msg;
```

That sends an empty-buttons-array inline keyboard, which Telegram interprets as "remove the keyboard from this message". No error, no leftover ghost keyboard.

### Alternatives, depending on what you actually want

- **Remove the keyboard but keep the message text/caption** — the snippet above. The text stays, the buttons go.
- **Replace text *and* remove keyboard in one call** — use `editMessageText`:
  ```js
  msg.payload = {
      type: "editMessageText",
      content: "New message text",
      options: {
          chat_id: msg.payload.chatId,
          message_id: msg.payload.messageId
          // Omit reply_markup entirely — Telegram drops the existing keyboard.
      }
  };
  ```
- **Delete the whole message** — use `deleteMessage`:
  ```js
  msg.payload = {
      type: "deleteMessage",
      chatId: msg.payload.chatId,
      messageId: msg.payload.messageId
  };
  ```

### Why your other variations failed

For the record (in case anyone else finds this thread):

| Variation | Why it failed |
|---|---|
| `reply_markup: null` in `options` | `content: "-"` was still being sent as the first positional arg → "object expected" |
| `reply_markup: ""` in `options` | Same — `content` is the actual reply-markup slot, the inner field is ignored |
| `reply_markup: '{"inline_keyboard":[]}'` (JSON **string**) in `options` | Same — and even if `content` were ignored, the lib expects an object, not a JSON-string |
| No `reply_markup` at all | Same — the empty `content: "-"` is still being passed as the markup |

The misleading part is that all four "succeeded" in stripping the keyboard from the message in the Telegram client — that's because Telegram processed the edit far enough to drop the markup before bailing on the response. The error came back over the wire, but the UI side-effect had already happened. So you saw both "it worked" and "it errored" together. Hence the confusion.

Closing as answered. If you hit the "object expected" error again on a different `editMessage*` call, the same logic applies — check what the first positional argument is supposed to be for that specific method.
